Dense -> CSR support batch dimensions

amjames opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82121
* #82122
* #83086
* #83085
* __->__ #83084

Only requires changes to the dense->sparse pathway. The reverse already has support. 


cc @alexsamardzic @nikitaved @pearu @cpuhrsch @bhosmer

amjames committed:
Dense -> CSR support batch dimensions

[ghstack-poisoned]

amjames committed:
Update on "Dense -> CSR support batch dimensions"


Only requires changes to the dense->sparse pathway. The reverse already has support. 


[ghstack-poisoned]

amjames commented on test/test_sparse_csr.py:
```diff
@@ -1208,14 +1208,6 @@ def test_csr_to_block_csr_errors(self, device, dtype):
             with self.assertRaisesRegex(RuntimeError, r"size \(16, 16\) with block size \(5, 5\)"):
                 block_t = t.to_sparse_bsr((5, 5))
 
-    @dtypes(*all_types_and_complex_and(torch.half, torch.bool, torch.bfloat16))
```
This test is obsolete. The error cases for `dense->sparse_csr` are now empty batch dims, and when batches do not have the same nnz. Both of these are covered by the new generalized `test_dense_to_from_sparse_compressed`

bhosmer reviewed (approved):
Nice!

amjames committed:
Update on "Dense -> CSR support batch dimensions"


Only requires changes to the dense->sparse pathway. The reverse already has support. 


[ghstack-poisoned]

amjames committed:
Update on "Dense -> CSR support batch dimensions"


Only requires changes to the dense->sparse pathway. The reverse already has support. 


[ghstack-poisoned]

amjames committed:
Update on "Dense -> CSR support batch dimensions"


Only requires changes to the dense->sparse pathway. The reverse already has support. 


[ghstack-poisoned]

amjames committed:
Update on "Dense -> CSR support batch dimensions"


Only requires changes to the dense->sparse pathway. The reverse already has support. 


[ghstack-poisoned]

amjames committed:
Update on "Dense -> CSR support batch dimensions"


Only requires changes to the dense->sparse pathway. The reverse already has support. 


[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -802,6 +792,47 @@ std::pair<Tensor, Tensor> _not_zero_mask_to_col_row_indices(
+      values.device());
+}
+
+Tensor dense_to_sparse_csc(const Tensor& self) {
+  return self.to_sparse().to_sparse_csc();
+}
```
Is this improved later in the stack? I mean, we can perform that without `to_sparse`, right?

nikitaved replied:
EDIT: this code is updated later in the stack.

amjames replied:
I'm not sure why this is detected as a change but the function was not added in the PR it just moved down because the function above is expanded. The exact same lines of code may be found in the deletions. I did not touch the `dense_to_sparse_csc` implementation in this PR

nikitaved reviewed (approved):
LGTM! Thank you, @amjames ! One super minor nit: this PR adds some code which is later removed in the follow-up PRs.

amjames committed:
Update on "Dense -> CSR support batch dimensions"


Only requires changes to the dense->sparse pathway. The reverse already has support. 


[ghstack-poisoned]

amjames committed:
Update on "Dense -> CSR support batch dimensions"


Only requires changes to the dense->sparse pathway. The reverse already has support. 


[ghstack-poisoned]

amjames commented:
@pytorchbot merge

amjames committed:
Update on "Dense -> CSR support batch dimensions"


Only requires changes to the dense->sparse pathway. The reverse already has support. 


[ghstack-poisoned]
